File: deno-debugger/scripts/heap_analyzer.ts
/**
 * Heap snapshot analysis for memory leak detection.
 *
 * Provides tools to:
 * - Parse V8 heap snapshots
 * - Compare snapshots to find memory growth
 * - Find retaining paths (why objects are kept alive)
 * - Detect leak patterns
 * - Analyze heap usage
 */

import type { HeapEdge, HeapNode, HeapSnapshotData, RetainingPath } from "./types.ts";
import type { CDPClient } from "./cdp_client.ts";

export interface NodeSizeSummary {
  nodeType: string;
  count: number;
  totalSize: number;
  avgSize: number;
}

export interface ComparisonRow {
  nodeType: string;
  name: string;
  countBefore: number;
  countAfter: number;
  countDelta: number;
  sizeBefore: number;
  sizeAfter: number;
  sizeDelta: number;
}

export interface LeakCandidate {
  nodeType: string;
  name: string;
  totalGrowthMB: number;
  growthPerSnapshotMB: number;
  snapshotsGrowing: number;
}

export interface LargestObject {
  nodeId: number;
  nodeType: string;
  name: string;
  sizeBytes: number;
  sizeMB: number;
}

export interface HeapSnapshotOptions {
  /**
   * Skip parsing edges (faster, but can't find retaining paths)
   * Use this when you only need node summaries for comparison
   * Default: false
   */
  skipEdges?: boolean;

  /**
   * Skip building retention index (faster, but can't find retaining paths)
   * Use this when you only need node data without relationships
   * Default: same as skipEdges
   */
  skipRetention?: boolean;
}

export class HeapSnapshot {
  public rawData: HeapSnapshotData;
  public snapshot: HeapSnapshotData["snapshot"];
  public nodes: HeapNode[] = [];
  public edges: HeapEdge[] = [];
  public strings: string[];
  public traceFunctionInfos: number[];
  public traceTree: number[];
  public nodeById: Map<number, HeapNode>;
  public retainedBy: Map<number, Array<[number, HeapEdge]>>;
  private options: HeapSnapshotOptions;

  constructor(snapshotData: HeapSnapshotData, options: HeapSnapshotOptions = {}) {
    this.options = {
      skipEdges: options.skipEdges ?? false,
      skipRetention: options.skipRetention ?? options.skipEdges ?? false,
    };

    this.rawData = snapshotData;
    this.snapshot = snapshotData.snapshot;
    this.strings = snapshotData.strings || [];
    this.traceFunctionInfos = snapshotData.trace_function_infos || [];
    this.traceTree = snapshotData.trace_tree || [];

    // Always parse nodes (needed for summaries)
    this.parseNodes();

    // Optionally parse edges (expensive for large heaps)
    if (!this.options.skipEdges) {
      this.parseEdges();
    }

    // Build indexes
    this.nodeById = new Map(this.nodes.map((n) => [n.id, n]));
    this.retainedBy = new Map();

    // Optionally build retention index (very expensive)
    if (!this.options.skipRetention && !this.options.skipEdges) {
      this.buildRetentionIndex();
    }
  }

  private parseNodes(): void {
    const meta = this.snapshot.meta;
    const nodeFields = meta.node_fields;
    const nodeTypes = meta.node_types[0];

    // Field indices
    const typeIdx = nodeFields.indexOf("type");
    const nameIdx = nodeFields.indexOf("name");
    const idIdx = nodeFields.indexOf("id");
    const selfSizeIdx = nodeFields.indexOf("self_size");
    const edgeCountIdx = nodeFields.indexOf("edge_count");
    const traceNodeIdIdx = nodeFields.indexOf("trace_node_id");

    const nodeFieldCount = nodeFields.length;
    const nodesData = this.rawData.nodes;

    for (let i = 0; i < nodesData.length; i += nodeFieldCount) {
      const node: HeapNode = {
        type: nodeTypes[nodesData[i + typeIdx]],
        name: this.strings[nodesData[i + nameIdx]],
        id: nodesData[i + idIdx],
        self_size: nodesData[i + selfSizeIdx],
        edge_count: nodesData[i + edgeCountIdx],
        trace_node_id: nodesData[i + traceNodeIdIdx],
        index: i,
      };
      this.nodes.push(node);
    }
  }

  private parseEdges(): void {
    const meta = this.snapshot.meta;
    const edgeFields = meta.edge_fields;
    const edgeTypes = meta.edge_types[0];

    // Field indices
    const typeIdx = edgeFields.indexOf("type");
    const nameOrIndexIdx = edgeFields.indexOf("name_or_index");
    const toNodeIdx = edgeFields.indexOf("to_node");

    const edgeFieldCount = edgeFields.length;
    const edgesData = this.rawData.edges;

    let fromNodeIndex = 0;

    for (let i = 0; i < edgesData.length; i += edgeFieldCount) {
      // Find which node this edge belongs to
      while (
        fromNodeIndex < this.nodes.length - 1 &&
        i >= this.calculateEdgeStartIndex(fromNodeIndex + 1)
      ) {
        fromNodeIndex++;
      }

      const edgeType = edgeTypes[edgesData[i + typeIdx]];
      let nameOrIndex: string | number = edgesData[i + nameOrIndexIdx];

      // Name or index depends on edge type
      if (edgeType === "property" || edgeType === "internal") {
        nameOrIndex = this.strings[nameOrIndex];
      }

      const edge: HeapEdge = {
        type: edgeType,
        name_or_index: nameOrIndex,
        to_node: edgesData[i + toNodeIdx],
        from_node: fromNodeIndex < this.nodes.length ? this.nodes[fromNodeIndex].id : 0,
      };
      this.edges.push(edge);
    }
  }

  private calculateEdgeStartIndex(nodeIndex: number): number {
    // Calculate where this node's edges start in the edges array
    let edgeCount = 0;
    for (let i = 0; i < nodeIndex && i < this.nodes.length; i++) {
      edgeCount += this.nodes[i].edge_count;
    }
    return edgeCount * this.snapshot.meta.edge_fields.length;
  }

  private buildRetentionIndex(): void {
    const nodeFieldCount = this.snapshot.meta.node_fields.length;

    let edgeIdx = 0;
    for (const node of this.nodes) {
      for (let i = 0; i < node.edge_count; i++) {
        if (edgeIdx < this.edges.length) {
          const edge = this.edges[edgeIdx];
          // Map target node -> (source node, edge)
          const targetNodeIdx = Math.floor(edge.to_node / nodeFieldCount);
          if (targetNodeIdx < this.nodes.length) {
            const targetNode = this.nodes[targetNodeIdx];
            if (!this.retainedBy.has(targetNode.id)) {
              this.retainedBy.set(targetNode.id, []);
            }
            this.retainedBy.get(targetNode.id)!.push([node.id, edge]);
          }
          edgeIdx++;
        }
      }
    }
  }

  getNodesByType(nodeType: string): HeapNode[] {
    return this.nodes.filter((n) => n.type === nodeType);
  }

  getNodesByName(name: string): HeapNode[] {
    return this.nodes.filter((n) => n.name === name);
  }

  getNodeSizeSummary(): NodeSizeSummary[] {
    const summary = new Map<string, { count: number; totalSize: number }>();

    for (const node of this.nodes) {
      const stats = summary.get(node.type) || { count: 0, totalSize: 0 };
      stats.count++;
      stats.totalSize += node.self_size;
      summary.set(node.type, stats);
    }

    const result: NodeSizeSummary[] = [];
    for (const [nodeType, stats] of summary.entries()) {
      result.push({
        nodeType,
        count: stats.count,
        totalSize: stats.totalSize,
        avgSize: stats.count > 0 ? stats.totalSize / stats.count : 0,
      });
    }

    return result.sort((a, b) => b.totalSize - a.totalSize);
  }

  findRetainingPath(nodeId: number, maxDepth = 10): RetainingPath | null {
    // Check if retention index was built
    if (this.options.skipRetention) {
      throw new Error(
        "Cannot find retaining paths: snapshot was created with skipRetention=true. " +
          "Create snapshot with { skipRetention: false } to enable this feature.",
      );
    }

    // BFS from node backwards to root
    const visited = new Set<number>();
    const queue: Array<[number, Array<{ node: HeapNode; edge: HeapEdge }>]> = [[nodeId, []]];

    while (queue.length > 0) {
      const [currentId, path] = queue.shift()!;

      if (visited.has(currentId) || path.length > maxDepth) {
        continue;
      }
      visited.add(currentId);

      const currentNode = this.nodeById.get(currentId);
      if (!currentNode) {
        continue;
      }

      // Check if this is a root node
      if (currentNode.type === "synthetic" && currentNode.name === "(GC roots)") {
        return {
          path,
          distance: path.length,
        };
      }

      // Follow retainers
      const retainers = this.retainedBy.get(currentId) || [];
      for (const [retainerId, edge] of retainers) {
        const retainerNode = this.nodeById.get(retainerId);
        if (retainerNode) {
          const newPath = [{ node: retainerNode, edge }, ...path];
          queue.push([retainerId, newPath]);
        }
      }
    }

    return null; // No path found
  }
}

export async function loadSnapshot(filePath: string): Promise<HeapSnapshot> {
  const data = JSON.parse(await Deno.readTextFile(filePath)) as HeapSnapshotData;
  return new HeapSnapshot(data);
}

export async function captureSnapshot(
  cdpClient: CDPClient,
  outputPath?: string,
): Promise<HeapSnapshot> {
  console.log("Capturing heap snapshot... (this may take a few seconds)");
  const snapshotJson = await cdpClient.takeHeapSnapshot();

  if (outputPath) {
    await Deno.writeTextFile(outputPath, snapshotJson);
    console.log(`Snapshot saved to ${outputPath}`);
  }

  const data = JSON.parse(snapshotJson) as HeapSnapshotData;
  return new HeapSnapshot(data);
}

export function compareSnapshots(before: HeapSnapshot, after: HeapSnapshot): ComparisonRow[] {
  // Build summaries by (type, name)
  function summarize(snapshot: HeapSnapshot) {
    const summary = new Map<string, { count: number; size: number }>();
    for (const node of snapshot.nodes) {
      const key = `${node.type}|${node.name}`;
      const stats = summary.get(key) || { count: 0, size: 0 };
      stats.count++;
      stats.size += node.self_size;
      summary.set(key, stats);
    }
    return summary;
  }

  const beforeSummary = summarize(before);
  const afterSummary = summarize(after);

  // Find all keys that exist in either snapshot
  const allKeys = new Set([...beforeSummary.keys(), ...afterSummary.keys()]);

  const data: ComparisonRow[] = [];
  for (const key of allKeys) {
    const [nodeType, name] = key.split("|");
    const beforeStats = beforeSummary.get(key) || { count: 0, size: 0 };
    const afterStats = afterSummary.get(key) || { count: 0, size: 0 };

    const countDelta = afterStats.count - beforeStats.count;
    const sizeDelta = afterStats.size - beforeStats.size;

    // Only include if there's actual growth
    if (countDelta > 0 || sizeDelta > 0) {
      data.push({
        nodeType,
        name,
        countBefore: beforeStats.count,
        countAfter: afterStats.count,
        countDelta,
        sizeBefore: beforeStats.size,
        sizeAfter: afterStats.size,
        sizeDelta,
      });
    }
  }

  return data.sort((a, b) => b.sizeDelta - a.sizeDelta);
}

/**
 * FAST PATH: Compare snapshots without building full HeapSnapshot objects
 * For large heaps (>100MB), this is 10-50x faster than compareSnapshots()
 *
 * Use this when:
 * - You only need comparison data (not retention paths)
 * - Heap snapshots are large (>100MB)
 * - You need quick results
 *
 * @param beforePath - Path to baseline snapshot file
 * @param afterPath - Path to comparison snapshot file
 * @returns Comparison results sorted by size delta
 */
export async function compareSnapshotsFast(
  beforePath: string,
  afterPath: string,
): Promise<ComparisonRow[]> {
  console.log("Fast comparison mode (skipping edges and retention paths)...\n");

  // Build summaries directly from raw data
  console.log("Processing baseline snapshot...");
  const beforeSummary = await buildSummaryFromFile(beforePath);

  console.log("Processing comparison snapshot...");
  const afterSummary = await buildSummaryFromFile(afterPath);

  console.log("Computing differences...\n");

  // Find all keys
  const allKeys = new Set([...beforeSummary.keys(), ...afterSummary.keys()]);

  const data: ComparisonRow[] = [];
  for (const key of allKeys) {
    const [nodeType, name] = key.split("|");
    const beforeStats = beforeSummary.get(key) || { count: 0, size: 0 };
    const afterStats = afterSummary.get(key) || { count: 0, size: 0 };

    const countDelta = afterStats.count - beforeStats.count;
    const sizeDelta = afterStats.size - beforeStats.size;

    if (countDelta > 0 || sizeDelta > 0) {
      data.push({
        nodeType,
        name,
        countBefore: beforeStats.count,
        countAfter: afterStats.count,
        countDelta,
        sizeBefore: beforeStats.size,
        sizeAfter: afterStats.size,
        sizeDelta,
      });
    }
  }

  return data.sort((a, b) => b.sizeDelta - a.sizeDelta);
}

/**
 * Build summary statistics directly from snapshot file
 * Skips creating HeapSnapshot instance (much faster for large heaps)
 */
async function buildSummaryFromFile(
  filePath: string,
): Promise<Map<string, { count: number; size: number }>> {
  const startTime = Date.now();
  const fileSize = (await Deno.stat(filePath)).size;
  console.log(`  File size: ${(fileSize / (1024 * 1024)).toFixed(1)} MB`);

  // Read and parse JSON
  console.log(`  Reading file...`);
  const jsonText = await Deno.readTextFile(filePath);

  console.log(`  Parsing JSON...`);
  const data = JSON.parse(jsonText) as HeapSnapshotData;

  console.log(`  Building summary...`);

  const summary = new Map<string, { count: number; size: number }>();

  const meta = data.snapshot.meta;
  const nodeFields = meta.node_fields;
  const nodeTypes = meta.node_types[0];
  const strings = data.strings || [];

  const typeIdx = nodeFields.indexOf("type");
  const nameIdx = nodeFields.indexOf("name");
  const selfSizeIdx = nodeFields.indexOf("self_size");
  const nodeFieldCount = nodeFields.length;

  const nodesData = data.nodes;
  const nodeCount = nodesData.length / nodeFieldCount;

  // Process in batches for progress feedback
  const batchSize = 100000;
  let processed = 0;

  for (let i = 0; i < nodesData.length; i += nodeFieldCount) {
    const type = nodeTypes[nodesData[i + typeIdx]];
    const name = strings[nodesData[i + nameIdx]];
    const selfSize = nodesData[i + selfSizeIdx];

    const key = `${type}|${name}`;
    const stats = summary.get(key);

    if (stats) {
      stats.count++;
      stats.size += selfSize;
    } else {
      summary.set(key, { count: 1, size: selfSize });
    }

    processed++;
    if (processed % batchSize === 0) {
      const pct = ((processed / nodeCount) * 100).toFixed(0);
      console.log(
        `    ${pct}% (${(processed / 1000).toFixed(0)}k / ${(nodeCount / 1000).toFixed(0)}k nodes)`,
      );
    }
  }

  const elapsed = ((Date.now() - startTime) / 1000).toFixed(1);
  console.log(`  ✓ Summary complete: ${summary.size} unique types in ${elapsed}s\n`);

  return summary;
}

export function detectLeaks(
  snapshots: HeapSnapshot[],
  thresholdMB = 1.0,
): LeakCandidate[] {
  if (snapshots.length < 2) {
    return [];
  }

  // Track growth trends
  const growthTrends = new Map<string, number[]>();

  for (let i = 0; i < snapshots.length - 1; i++) {
    const comparison = compareSnapshots(snapshots[i], snapshots[i + 1]);
    for (const row of comparison) {
      const key = `${row.nodeType}|${row.name}`;
      const deltas = growthTrends.get(key) || [];
      deltas.push(row.sizeDelta);
      growthTrends.set(key, deltas);
    }
  }

  // Find objects with consistent growth
  const leaks: LeakCandidate[] = [];
  const thresholdBytes = thresholdMB * 1024 * 1024;

  for (const [key, deltas] of growthTrends.entries()) {
    // Check if consistently growing
    if (deltas.every((d) => d > 0)) {
      const totalGrowth = deltas.reduce((sum, d) => sum + d, 0);
      if (totalGrowth > thresholdBytes) {
        const [nodeType, name] = key.split("|");
        leaks.push({
          nodeType,
          name,
          totalGrowthMB: totalGrowth / (1024 * 1024),
          growthPerSnapshotMB: (totalGrowth / deltas.length) / (1024 * 1024),
          snapshotsGrowing: deltas.length,
        });
      }
    }
  }

  return leaks.sort((a, b) => b.totalGrowthMB - a.totalGrowthMB);
}

export function findLargestObjects(snapshot: HeapSnapshot, limit = 20): LargestObject[] {
  const data: LargestObject[] = [];

  for (const node of snapshot.nodes) {
    if (node.self_size > 0) {
      data.push({
        nodeId: node.id,
        nodeType: node.type,
        name: node.name,
        sizeBytes: node.self_size,
        sizeMB: node.self_size / (1024 * 1024),
      });
    }
  }

  return data.sort((a, b) => b.sizeBytes - a.sizeBytes).slice(0, limit);
}

// ============================================================================
// CLI Usage
// ============================================================================

if (import.meta.main) {
  console.log("Heap Analyzer - Example Usage");
  console.log("==============================");
  console.log();
  console.log("// Load snapshot from file");
  console.log("const snapshot = await loadSnapshot('heap.heapsnapshot');");
  console.log();
  console.log("// Get size summary by type");
  console.log("const summary = snapshot.getNodeSizeSummary();");
  console.log("console.table(summary);");
  console.log();
  console.log("// Find largest objects");
  console.log("const large = findLargestObjects(snapshot);");
  console.log("console.table(large);");
  console.log();
  console.log("// Compare two snapshots");
  console.log("const before = await loadSnapshot('before.heapsnapshot');");
  console.log("const after = await loadSnapshot('after.heapsnapshot');");
  console.log("const growth = compareSnapshots(before, after);");
  console.log("console.table(growth.slice(0, 10));");
  console.log();
  console.log("// Detect leaks over time");
  console.log("const snapshots = await Promise.all([");
  console.log("  loadSnapshot('snap_0.heapsnapshot'),");
  console.log("  loadSnapshot('snap_1.heapsnapshot'),");
  console.log("  loadSnapshot('snap_2.heapsnapshot'),");
  console.log("]);");
  console.log("const leaks = detectLeaks(snapshots);");
  console.log("console.table(leaks);");
}

File: deno-debugger/scripts/heap_analyzer.py
"""
Heap snapshot analysis for memory leak detection.

Provides tools to:
- Capture and parse V8 heap snapshots
- Compare snapshots to find memory growth
- Find retaining paths (why objects are kept alive)
- Detect leak patterns
- Generate pandas DataFrames for analysis
"""

import json
from collections import defaultdict
from dataclasses import dataclass
from typing import Any, Dict, List, Optional, Tuple

import pandas as pd


@dataclass
class HeapNode:
    """Represents a node in the heap snapshot."""

    node_id: int
    node_type: str
    name: str
    self_size: int
    edge_count: int
    trace_node_id: int


@dataclass
class HeapEdge:
    """Represents an edge (reference) between heap nodes."""

    edge_type: str
    name_or_index: Any
    to_node: int


class HeapSnapshot:
    """
    Parsed V8 heap snapshot with analysis capabilities.

    V8 heap snapshots use a compact format with parallel arrays.
    This class parses and indexes the snapshot for efficient queries.
    """

    def __init__(self, snapshot_data: Dict[str, Any]):
        self.raw_data = snapshot_data
        self.snapshot = snapshot_data.get("snapshot", {})
        self.nodes = []
        self.edges = []
        self.strings = snapshot_data.get("strings", [])
        self.trace_function_infos = snapshot_data.get("trace_function_infos", [])
        self.trace_tree = snapshot_data.get("trace_tree", [])

        # Parse nodes and edges
        self._parse_nodes()
        self._parse_edges()

        # Build indexes for fast lookup
        self.node_by_id = {node.node_id: node for node in self.nodes}
        self._build_retention_index()

    def _parse_nodes(self):
        """Parse the nodes array into HeapNode objects."""
        meta = self.snapshot.get("meta", {})
        node_fields = meta.get("node_fields", [])
        node_types = meta.get("node_types", [[]])[0]

        # Field indices
        type_idx = node_fields.index("type")
        name_idx = node_fields.index("name")
        id_idx = node_fields.index("id")
        self_size_idx = node_fields.index("self_size")
        edge_count_idx = node_fields.index("edge_count")
        trace_node_id_idx = node_fields.index("trace_node_id")

        node_field_count = len(node_fields)
        nodes_data = self.raw_data.get("nodes", [])

        for i in range(0, len(nodes_data), node_field_count):
            node_data = nodes_data[i : i + node_field_count]

            node = HeapNode(
                node_id=node_data[id_idx],
                node_type=node_types[node_data[type_idx]],
                name=self.strings[node_data[name_idx]],
                self_size=node_data[self_size_idx],
                edge_count=node_data[edge_count_idx],
                trace_node_id=node_data[trace_node_id_idx],
            )
            self.nodes.append(node)

    def _parse_edges(self):
        """Parse the edges array into HeapEdge objects."""
        meta = self.snapshot.get("meta", {})
        edge_fields = meta.get("edge_fields", [])
        edge_types = meta.get("edge_types", [[]])[0]

        # Field indices
        type_idx = edge_fields.index("type")
        name_or_index_idx = edge_fields.index("name_or_index")
        to_node_idx = edge_fields.index("to_node")

        edge_field_count = len(edge_fields)
        edges_data = self.raw_data.get("edges", [])

        for i in range(0, len(edges_data), edge_field_count):
            edge_data = edges_data[i : i + edge_field_count]

            # Name or index depends on edge type
            name_or_index = edge_data[name_or_index_idx]
            edge_type = edge_types[edge_data[type_idx]]

            if edge_type in ["property", "internal"]:
                name_or_index = self.strings[name_or_index]

            edge = HeapEdge(
                edge_type=edge_type,
                name_or_index=name_or_index,
                to_node=edge_data[to_node_idx],
            )
            self.edges.append(edge)

    def _build_retention_index(self):
        """Build index of what references each node (for retaining paths)."""
        self.retained_by = defaultdict(list)

        edge_idx = 0
        for node in self.nodes:
            for _ in range(node.edge_count):
                if edge_idx < len(self.edges):
                    edge = self.edges[edge_idx]
                    # Map target node -> (source node, edge)
                    target_node_idx = edge.to_node // len(
                        self.snapshot.get("meta", {}).get("node_fields", [])
                    )
                    if target_node_idx < len(self.nodes):
                        self.retained_by[self.nodes[target_node_idx].node_id].append(
                            (node.node_id, edge)
                        )
                    edge_idx += 1

    def get_nodes_by_type(self, node_type: str) -> List[HeapNode]:
        """Get all nodes of a specific type."""
        return [node for node in self.nodes if node.node_type == node_type]

    def get_nodes_by_name(self, name: str) -> List[HeapNode]:
        """Get all nodes with a specific name."""
        return [node for node in self.nodes if node.name == name]

    def get_node_size_summary(self) -> pd.DataFrame:
        """
        Get summary of heap usage by node type.

        Returns:
            DataFrame with columns: node_type, count, total_size, avg_size
        """
        summary = defaultdict(lambda: {"count": 0, "total_size": 0})

        for node in self.nodes:
            summary[node.node_type]["count"] += 1
            summary[node.node_type]["total_size"] += node.self_size

        data = []
        for node_type, stats in summary.items():
            data.append(
                {
                    "node_type": node_type,
                    "count": stats["count"],
                    "total_size": stats["total_size"],
                    "avg_size": (
                        stats["total_size"] / stats["count"]
                        if stats["count"] > 0
                        else 0
                    ),
                }
            )

        df = pd.DataFrame(data)
        return df.sort_values("total_size", ascending=False)

    def find_retaining_path(
        self, node_id: int, max_depth: int = 10
    ) -> List[Tuple[HeapNode, HeapEdge]]:
        """
        Find retaining path from root to a node (why it's alive).

        Returns:
            List of (node, edge) tuples showing the path
        """
        # BFS from node backwards to root
        visited = set()
        queue = [(node_id, [])]

        while queue:
            current_id, path = queue.pop(0)

            if current_id in visited or len(path) > max_depth:
                continue
            visited.add(current_id)

            current_node = self.node_by_id.get(current_id)
            if not current_node:
                continue

            # Check if this is a root node
            if (
                current_node.node_type == "synthetic"
                and current_node.name == "(GC roots)"
            ):
                return path

            # Follow retainers
            for retainer_id, edge in self.retained_by.get(current_id, []):
                retainer_node = self.node_by_id.get(retainer_id)
                if retainer_node:
                    new_path = [(retainer_node, edge)] + path
                    queue.append((retainer_id, new_path))

        return []  # No path found


def load_snapshot(file_path: str) -> HeapSnapshot:
    """
    Load a heap snapshot from file.

    Args:
        file_path: Path to .heapsnapshot JSON file

    Returns:
        Parsed HeapSnapshot object
    """
    with open(file_path, "r") as f:
        data = json.load(f)
    return HeapSnapshot(data)


async def capture_snapshot(
    cdp_client, output_path: Optional[str] = None
) -> HeapSnapshot:
    """
    Capture a heap snapshot from a connected CDP client.

    Args:
        cdp_client: Connected CDPClient instance
        output_path: Optional path to save snapshot JSON

    Returns:
        Parsed HeapSnapshot object
    """
    print("Capturing heap snapshot... (this may take a few seconds)")
    snapshot_json = await cdp_client.take_heap_snapshot()

    if output_path:
        with open(output_path, "w") as f:
            f.write(snapshot_json)
        print(f"Snapshot saved to {output_path}")

    data = json.loads(snapshot_json)
    return HeapSnapshot(data)


def compare_snapshots(before: HeapSnapshot, after: HeapSnapshot) -> pd.DataFrame:
    """
    Compare two snapshots to find memory growth.

    Args:
        before: Earlier snapshot
        after: Later snapshot

    Returns:
        DataFrame showing objects that grew, with columns:
        - node_type: Type of object
        - name: Object name/constructor
        - count_before: Count in before snapshot
        - count_after: Count in after snapshot
        - count_delta: Change in count
        - size_before: Total size before
        - size_after: Total size after
        - size_delta: Change in size
    """

    # Build summaries by (type, name)
    def summarize(snapshot):
        summary = defaultdict(lambda: {"count": 0, "size": 0})
        for node in snapshot.nodes:
            key = (node.node_type, node.name)
            summary[key]["count"] += 1
            summary[key]["size"] += node.self_size
        return summary

    before_summary = summarize(before)
    after_summary = summarize(after)

    # Find all keys that exist in either snapshot
    all_keys = set(before_summary.keys()) | set(after_summary.keys())

    data = []
    for key in all_keys:
        node_type, name = key
        before_stats = before_summary.get(key, {"count": 0, "size": 0})
        after_stats = after_summary.get(key, {"count": 0, "size": 0})

        count_delta = after_stats["count"] - before_stats["count"]
        size_delta = after_stats["size"] - before_stats["size"]

        # Only include if there's actual growth
        if count_delta > 0 or size_delta > 0:
            data.append(
                {
                    "node_type": node_type,
                    "name": name,
                    "count_before": before_stats["count"],
                    "count_after": after_stats["count"],
                    "count_delta": count_delta,
                    "size_before": before_stats["size"],
                    "size_after": after_stats["size"],
                    "size_delta": size_delta,
                }
            )

    df = pd.DataFrame(data)
    if not df.empty:
        df = df.sort_values("size_delta", ascending=False)

    return df


def detect_leaks(
    snapshots: List[HeapSnapshot], threshold_mb: float = 1.0
) -> pd.DataFrame:
    """
    Detect potential memory leaks from a series of snapshots.

    Looks for objects that grow consistently across snapshots.

    Args:
        snapshots: List of snapshots taken over time
        threshold_mb: Only report objects growing more than this (in MB)

    Returns:
        DataFrame of likely leaking objects
    """
    if len(snapshots) < 2:
        return pd.DataFrame()

    # Track growth trends
    growth_trends = defaultdict(list)

    for i in range(len(snapshots) - 1):
        comparison = compare_snapshots(snapshots[i], snapshots[i + 1])
        for _, row in comparison.iterrows():
            key = (row["node_type"], row["name"])
            growth_trends[key].append(row["size_delta"])

    # Find objects with consistent growth
    leaks = []
    threshold_bytes = threshold_mb * 1024 * 1024

    for key, deltas in growth_trends.items():
        # Check if consistently growing
        if all(d > 0 for d in deltas):
            total_growth = sum(deltas)
            if total_growth > threshold_bytes:
                node_type, name = key
                leaks.append(
                    {
                        "node_type": node_type,
                        "name": name,
                        "total_growth_mb": total_growth / (1024 * 1024),
                        "growth_per_snapshot_mb": (total_growth / len(deltas))
                        / (1024 * 1024),
                        "snapshots_growing": len(deltas),
                    }
                )

    df = pd.DataFrame(leaks)
    if not df.empty:
        df = df.sort_values("total_growth_mb", ascending=False)

    return df


def find_largest_objects(snapshot: HeapSnapshot, limit: int = 20) -> pd.DataFrame:
    """
    Find the largest objects in a snapshot.

    Args:
        snapshot: Heap snapshot
        limit: Number of objects to return

    Returns:
        DataFrame of largest objects
    """
    data = []
    for node in snapshot.nodes:
        if node.self_size > 0:
            data.append(
                {
                    "node_id": node.node_id,
                    "node_type": node.node_type,
                    "name": node.name,
                    "size_bytes": node.self_size,
                    "size_mb": node.self_size / (1024 * 1024),
                }
            )

    df = pd.DataFrame(data)
    if not df.empty:
        df = df.sort_values("size_bytes", ascending=False).head(limit)

    return df


if __name__ == "__main__":
    print("Heap Analyzer - Example Usage")
    print("==============================")
    print()
    print("# Load snapshot from file")
    print("snapshot = load_snapshot('heap.heapsnapshot')")
    print()
    print("# Get size summary by type")
    print("summary = snapshot.get_node_size_summary()")
    print("print(summary)")
    print()
    print("# Find largest objects")
    print("large = find_largest_objects(snapshot)")
    print("print(large)")
    print()
    print("# Compare two snapshots")
    print("before = load_snapshot('before.heapsnapshot')")
    print("after = load_snapshot('after.heapsnapshot')")
    print("growth = compare_snapshots(before, after)")
    print("print(growth.head(10))")
    print()
    print("# Detect leaks over time")
    print("snapshots = [load_snapshot(f'snap_{i}.heapsnapshot') for i in range(5)]")
    print("leaks = detect_leaks(snapshots)")
    print("print(leaks)")
